engine/src/silo/whitelist.rs

```diff
@@ -0,0 +1,156 @@
+use aurora_engine_sdk::io::{StorageIntermediate, IO};
+use aurora_engine_types::storage::{bytes_to_key, KeyPrefix};
+use aurora_engine_types::AsBytes;
+use borsh::{BorshDeserialize, BorshSerialize};
+
+use crate::prelude::Vec;
+use crate::silo::parameters::{WhitelistKindArgs, WhitelistStatusArgs};
+
+const STATUS: &[u8] = b"LIST_STATUS";
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, BorshSerialize, BorshDeserialize)]
+pub enum WhitelistKind {
+    Admin = 0x0,
+    EvmAdmin = 0x1,
+    Account = 0x2,
+    Address = 0x3,
+}
+
+impl From<WhitelistKind> for u8 {
+    fn from(list: WhitelistKind) -> Self {
+        match list {
+            WhitelistKind::Admin => 0x0,
+            WhitelistKind::EvmAdmin => 0x1,
+            WhitelistKind::Account => 0x2,
+            WhitelistKind::Address => 0x3,
+        }
+    }
+}
+
+/// `WhiteList` for checking access.
+/// * io - I/O trait handler
+pub struct WhiteList<I> {
+    io: I,
+    kind: WhitelistKind,
+}
+
+impl<I> WhiteList<I>
+where
+    I: IO + Copy,
+{
+    pub const fn init(io: &I, kind: WhitelistKind) -> Self {
+        Self { io: *io, kind }
+    }
+
+    pub fn enable(&mut self) {
+        let key = self.key(STATUS);
+        self.io.write_storage(&key, &[1]);
+    }
+
+    pub fn disable(&mut self) {
+        let key = self.key(STATUS);
+        self.io.write_storage(&key, &[0]);
+    }
+
+    pub fn is_enabled(&self) -> bool {
+        // White list is enabled by default. So return `true` if the key doesn't exist.
+        let key = self.key(STATUS);
+        self.io
+            .read_storage(&key)
+            .map_or(true, |value| value.to_vec() == [1])
```

Comment:
nit: Does it make sense to allocate a vector here?

Reply:
Yes. It does. In such a way, we provide a place in the memory via mutable reference to allocated vector for writing bytes that we read from the near storage.

Reply:
I meant, can we avoid a dynamic allocation just for a comparison? It's probably not a big deal but I'm not sure how much dynamic allocations affect gas usage.

Reply:
I understand your question and concern 😄 . The short answer is No. We can't.